hash/hashserializer/hashserializer.go

```diff
@@ -0,0 +1,181 @@
+//
+// The following values are all considered as empty and will not be part of the generated string:
+//  string("")
+//  bool(false)
+//  int(0)
+//  float(0.0)
+//  []string{}
+//  []string{""}
+//  map[string]string{}
+//  map[string]string{"a":"", "b":""}
+//
+// String
+//
+// String values are the simplest are the value doesn't need to be encoded.
+//
+// For prefix "foo" and value "bar", the result is:
+//  foo:bar;
+//
+// Bool
+//
+// Only bool value true produce an output as the default value false is discarded.
+//
+// For prefix "foo" and value "true", the result is:
+//  foo:true;
+//
+// Int
+//
+// Int values are encoded to string using base 10 representation. If the value is negative, the sign "-" should prefix the value.
+//
+// For prefix "foo" and value "42", the result is:
+//  foo:42;
+//
+// Float
+//
+// Float values are encoded to string using base 10 with no exponent. If the value is negative, the sign "-" should prefix the value.
+//
+// For prefix "foo" and value "3.14159265359", the result is:
+//  foo:3.14159265359;
+//
+// Slice
+//
+// For slices, prefixes are the index of the elements. Empty value are discarded.
+//
+//  []string{"foo", "", "bar"}
+//  0:foo;2:bar;
+//
+// Map
+//
+// For maps, prefixes are the keys. The maps must be sorted by keys to get a deterministic result. Empty value are discarded.
+//
+//  map[string]string{"c":"bar", "b": "", "a":"foo"}
+//  a:foo;c:bar;
+//
+// Struct
+//
+// Struct should implement the function `HashSerialize() string` to return the hash-serialized string.
+//
+// Nested
+//
+// When adding a struct, a slice or a map, the value is hash-serialized before added to the string.
+//
+// For prefix "foo" and value `[]string{"foo", "", "bar"}`, the result is:
+//  foo:0:foo;2:bar;;
+//
+//
+package hashserializer
+
+import (
+	"bytes"
+	"strconv"
+)
+
+// HashSerializable represents a serializable struct.
+type HashSerializable interface {
+	HashSerialize() string
+}
+
+// HashSerializer helps to create hash-serialized string of a struct.
+type HashSerializer struct {
+	buf bytes.Buffer
+}
+
+// New returns a new HashSerializer.
+func New() *HashSerializer {
+	return &HashSerializer{}
+}
+
+// HashSerialize returns the hash-serialized constructed string.
+func (s *HashSerializer) HashSerialize() string {
+	return s.buf.String()
+}
+
+// AddString adds a string to the serializer.
+// If value is empty, nothing is added to the HashSerializer.
+func (s *HashSerializer) AddString(prefix string, value string) {
+	if value != "" {
+		s.buf.WriteString(prefix)
+		s.buf.WriteString(":")
+		s.buf.WriteString(value)
+		s.buf.WriteString(";")
+	}
+}
+
+// AddBool adds a boolean to the serializer.
+// If value is false, nothing is added to the HashSerializer.
+func (s *HashSerializer) AddBool(prefix string, value bool) {
```

Comment:
These `AddX` method could be chainable if they return themselves  (not blocking at all)

```go
func (s *HashSerializer) AddBool(prefix string, value bool) *HashSerializer {
  if value {
	s.AddString(prefix, "true")
  }
  return s
}
```

with this we could use it like
```go
New().
  AddBool(1, true).
  AddString(2, "ttt").
  HashSerialize()
```